prototype/southbound/device-factory/device-factory-core/src/main/java/org/eclipse/sensinact/gateway/southbound/device/factory/impl/FactoryParserHandler.java

```diff
@@ -0,0 +1,709 @@
+        }
+
+        return newPlaceholders;
+    }
+
+    /**
+     * Assigns a value to each variable or throws an exception
+     *
+     * @param configuration     Mapping configuration
+     * @param record            Current record
+     * @param variablesMappings Definitions of variables
+     * @return
+     * @throws ParserException Error resolving variables
+     */
+    private Map<String, String> resolveVariables(final DeviceMappingConfigurationDTO configuration,
+            final IDeviceMappingRecord record, final Map<String, RecordPath> variablesMappings) throws ParserException {
+
+        Set<String> previouslyRemaining = new HashSet<>();
+        final Set<String> remainingVars = new HashSet<>(variablesMappings.keySet());
+        final Map<String, String> resolvedVars = new HashMap<>();
+
+        while (!remainingVars.isEmpty()) {
+            if (previouslyRemaining.equals(remainingVars)) {
+                throw new ParserException("Can't resolve variables: " + remainingVars);
+            }
+
+            previouslyRemaining = Set.copyOf(remainingVars);
+            final Set<String> resolved = new HashSet<>();
+            for (final String varName : remainingVars) {
+                final String path = String.valueOf(variablesMappings.get(varName));
+                final Matcher matcher = VariableSolver.varPattern.matcher(path);
+                if (matcher.find()) {
+                    String newPath = path;
+                    boolean fullyResolved = true;
+                    do {
+                        final String innerVar = matcher.group(1);
+                        final Object resolvedPath = resolvedVars.get("$" + innerVar);
+                        if (resolvedPath != null) {
+                            newPath = newPath.replace("${" + innerVar + "}", String.valueOf(resolvedPath));
+                        } else {
+                            // Not yet available
+                            fullyResolved = false;
+                            break;
+                        }
+
+                        if (fullyResolved) {
+                            resolvedVars.put(varName, newPath);
+                            resolved.add(varName);
+                        }
+                    } while (matcher.find());
+                } else {
+                    // Direct field access
+                    resolvedVars.put(varName, (String) record.getField(new RecordPath(path)));
+                    resolved.add(varName);
+                }
+            }
+
+            remainingVars.removeAll(resolved);
+        }
+
+        return resolvedVars;
+    }
+
+    /**
+     * Looks for a location in the given record
+     *
+     * @param record       Record to read
+     * @param placeholders Defined mapping placeholders
+     * @return The parsed location as a GeoJSON string or null
+     */
+    private String computeLocation(final IDeviceMappingRecord record, final Map<String, RecordPath> placeholders) {
```

Comment:
I think that this should be returning a GeoJson Object, not a String. That's what the value of the admin/location should be.

Reply:
Agreed, will do